osu.Game/Database/EFToRealmMigrator.cs

```diff
@@ -209,6 +218,8 @@ private void beginMigration()
                 // If we were to not do this, the migration would run another time the next time the user starts the game.
                 deletePreRealmData();
 
+                realmBlockOperations.Dispose();
```

Comment:
This call is not safe when combined with passing `blockAllOperations` to `CreateBackup()`, because:

* Blocking operation is started on line 131
* In line 135, the blocking operation is passed to `CreateBackup()`, which uses it in a `using` block. Therefore, once that `using` block terminates, the blocking operation disposable is disposed once.
* Then, here in the continuation, the blocking operation is disposed a second time.

`InvokeOnDisposal` doesn't check if the action it was given was already invoked once, so it will potentially be invoked multiple times. `BlockAllOperations()` uses the following callback:

https://github.com/ppy/osu/blob/c7287556e5cc42ab2d97bf5143dff3cabd7c20ba/osu.Game/Database/RealmAccess.cs#L881-L893

The main issue here is `realmRetrievalLock`, which is a semaphore with no max capacity. Therefore, the disposable is *not* idempotent - multiple disposes will release multiple items to the semaphore, which means that subsequent waits on this semaphore will most likely never block anything, because the expectation is that this semaphore should essentially act as a mutex, and it cannot do so if it has been released more than once.

I've confirmed that this is a real scenario in debug, and it could lead to things like storage relocations failing (because `BlockAllOperations()` will cease to work).

Reply:
Interesting. I tested as far as releasing twice, but not attempting to block again afterwards.

Any simple solution to avoiding that?

Reply:
The simplest thing would probably be to make `CreateBackup()` accept a boolean instead that indicates if it should take a `BlockAllOperations()` disposable. Setting max count on the semaphore is out of the question as it will throw if you attempt to release more items to the semaphore than the max count.

Reply:
I could agree to that too, sure - then again, there's also the question as to whether realm should be unblocked immediately after the backup is taken, or only after the full migration is completed. This code as written is doing the former and I don't immediately see a huge issue with that, but if that's viewed as undesirable, then such a solution won't work.

Reply:
I don't think it can block for the full duration. That means the realm calls won't work. The only reason fro the final `Dispose` is in case there's an error.

Reply:
Yeah I don't think guaranteeing `InvokeOnDisposal` to execute once would be good unless it's fine to dispose after creating backup. Would imagine solving this simply by:
```csharp
        public void CreateBackup(string backupFilename, IDisposable? blockAllOperations = null)
        {
            if (blockAllOperations != null)
                createBackup(backupFilename);
            else
            {
                using (BlockAllOperations())
                    createBackup(backupFilename);
            }
        }
```